fvm/src/kernel/default.rs

```diff
@@ -1006,41 +1006,99 @@ impl<C> EventOps for DefaultKernel<C>
+            );
         }
 
-        let actor_evt = {
-            let res = match panic::catch_unwind(|| {
-                fvm_ipld_encoding::from_slice(raw_evt).or_error(ErrorNumber::IllegalArgument)
-            }) {
-                Ok(v) => v,
-                Err(e) => {
-                    log::error!("panic when decoding event cbor from actor: {:?}", e);
-                    Err(syscall_error!(IllegalArgument; "panic when decoding event cbor from actor").into())
-                }
+        let total_buf_len = raw_evt.len() + raw_key.len() + raw_val.len();
+
+        let _t = self.call_manager.charge_gas(
+            self.call_manager
+                .price_list()
+                .on_actor_event_validate(total_buf_len),
+        )?;
+
+        const BYTES_PER_ENTRY: usize = 24;
+        if raw_evt.len() % BYTES_PER_ENTRY != 0 {
+            return Err(syscall_error!(IllegalArgument; "event not divisible by 24 bytes").into());
+        }
+        let nr_entries = raw_evt.len() / BYTES_PER_ENTRY;
+
+        if nr_entries > MAX_NR_ENTRIES {
+            return Err(syscall_error!(IllegalArgument; "event exceeded max entries: {} > {MAX_NR_ENTRIES}", nr_entries).into());
+        }
+
+        let mut key_offset: usize = 0;
+        let mut val_offset: usize = 0;
+
+        let mut entries: Vec<Entry> = Vec::with_capacity(nr_entries);
+        for i in 0..nr_entries {
+            let offset = i * BYTES_PER_ENTRY;
+            let view = &raw_evt[offset..offset + BYTES_PER_ENTRY];
+
+            // parse the fixed sized fields from the raw_evt buffer
+            //
+            let entry_fixed: EntryFixed = unsafe {
+                std::mem::transmute::<[u8; BYTES_PER_ENTRY], EntryFixed>(view.try_into().unwrap())
+            };
+
+            // make sure that the fixed parsed values are within bounds before we do any allocation
+            //
+            if entry_fixed.key_len > MAX_KEY_LEN as u32 {
+                return Err(syscall_error!(IllegalArgument; "event key exceeded max size: {} > {MAX_KEY_LEN}", entry_fixed.key_len  as u32).into());
+            }
+            if entry_fixed.val_len > MAX_VALUE_LEN as u32 {
+                return Err(syscall_error!(IllegalArgument; "event value exceeded max size: {} > {MAX_VALUE_LEN}", entry_fixed.val_len as u32).into());
+            }
+            if entry_fixed.codec != IPLD_RAW {
+                return Err(
+                    syscall_error!(IllegalCodec; "event codec must be IPLD_RAW, was: {}",  entry_fixed.codec as u64)
+                        .into(),
+                );
+            }
+
+            // parse the variable sized fields from the raw_key/raw_val buffers
+            //
+            let key = match std::str::from_utf8(
+                &raw_key[key_offset..key_offset + entry_fixed.key_len as usize],
+            ) {
+                Ok(s) => s,
+                Err(_) => return Err(syscall_error!(IllegalArgument, "event key").into()),
```

Comment:
```suggestion
                Err(e) => return Err(syscall_error!(IllegalArgument, "invalid event key: {}", e).into()),
```

(or something like that)